// angular4-paystack/esm2020/lib/angular4-paystack.directive.mjs
import { Directive, Input, Output, EventEmitter, HostListener } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./angular4-paystack.service";
export class Angular4PaystackDirective {
    constructor(paystackService) {
        this.paystackService = paystackService;
        this.paymentInit = new EventEmitter(); // tslint:disable-line
        this.onClose = new EventEmitter(); // tslint:disable-line
        this.callback = new EventEmitter();
        this.isPaying = false;
    }
    async pay() {
        let errorText = '';
        if (this.paystackOptions && Object.keys(this.paystackOptions).length >= 2) {
            errorText = this.valdateInput(this.paystackOptions);
            this.generateOptions(this.paystackOptions);
        }
        else {
            errorText = this.valdateInput(this);
            this.generateOptions(this);
        }
        if (errorText) {
            console.error(errorText);
            return errorText;
        }
        await this.paystackService.loadScript();
        if (this.isPaying) {
            return;
        }
        if (this.paymentInit.observers.length) {
            this.paymentInit.emit();
        }
        const payment = window.PaystackPop.setup(this._paystackOptions);
        payment.openIframe();
        this.isPaying = true;
        return '';
    }
    valdateInput(obj) {
        if (!this.callback.observers.length) {
            return 'ANGULAR-PAYSTACK: Insert a callback output like so (callback)=\'PaymentComplete($event)\' to check payment status';
        }
        return this.paystackService.checkInput(obj);
    }
    generateOptions(obj) {
        this._paystackOptions = this.paystackService.getPaystackOptions(obj);
        this._paystackOptions.onClose = () => {
            if (this.onClose.observers.length) {
                this.isPaying = false;
                this.onClose.emit();
            }
        };
        this._paystackOptions.callback = (...response) => {
            this.isPaying = false;
            this.callback.emit(...response);
        };
    }
    async buttonClick() {
        this.pay();
    }
}
Angular4PaystackDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.1.4", ngImport: i0, type: Angular4PaystackDirective, deps: [{ token: i1.Angular4PaystackService }], target: i0.ɵɵFactoryTarget.Directive });
Angular4PaystackDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.1.4", type: Angular4PaystackDirective, selector: "[angular4-paystack]", inputs: { key: "key", email: "email", amount: "amount", metadata: "metadata", ref: "ref", currency: "currency", plan: "plan", quantity: "quantity", subaccount: "subaccount", channels: "channels", transaction_charge: "transaction_charge", bearer: "bearer", class: "class", style: "style", paystackOptions: "paystackOptions" }, outputs: { paymentInit: "paymentInit", onClose: "onClose", callback: "callback" }, host: { listeners: { "click": "buttonClick()" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.1.4", ngImport: i0, type: Angular4PaystackDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[angular4-paystack]', // tslint:disable-line
                }]
        }], ctorParameters: function () { return [{ type: i1.Angular4PaystackService }]; }, propDecorators: { key: [{
                type: Input
            }], email: [{
                type: Input
            }], amount: [{
                type: Input
            }], metadata: [{
                type: Input
            }], ref: [{
                type: Input
            }], currency: [{
                type: Input
            }], plan: [{
                type: Input
            }], quantity: [{
                type: Input
            }], subaccount: [{
                type: Input
            }], channels: [{
                type: Input
            }], transaction_charge: [{
                type: Input
            }], bearer: [{
                type: Input
            }], class: [{
                type: Input
            }], style: [{
                type: Input
            }], paystackOptions: [{
                type: Input
            }], paymentInit: [{
                type: Output
            }], onClose: [{
                type: Output
            }], callback: [{
                type: Output
            }], buttonClick: [{
                type: HostListener,
                args: ['click']
            }] } });
