// angular4-paystack/esm2020/lib/angular4-paystack.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Angular4PaystackComponent } from './angular4-paystack.component';
import { Angular4PaystackEmbedComponent } from './angular4-paystack-embed.component';
import { Angular4PaystackDirective } from './angular4-paystack.directive';
import { Angular4PaystackService } from './angular4-paystack.service';
import { PUBLIC_KEY_TOKEN } from './paystack-token';
import * as i0 from "@angular/core";
export class Angular4PaystackModule {
    static forRoot(token) {
        return {
            ngModule: Angular4PaystackModule,
            providers: [
                Angular4PaystackService,
                { provide: PUBLIC_KEY_TOKEN, useValue: token }
            ]
        };
    }
}
Angular4PaystackModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.1.4", ngImport: i0, type: Angular4PaystackModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
Angular4PaystackModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.1.4", ngImport: i0, type: Angular4PaystackModule, declarations: [Angular4PaystackComponent, Angular4PaystackDirective, Angular4PaystackEmbedComponent], imports: [CommonModule], exports: [Angular4PaystackComponent, Angular4PaystackDirective, Angular4PaystackEmbedComponent] });
Angular4PaystackModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.1.4", ngImport: i0, type: Angular4PaystackModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.1.4", ngImport: i0, type: Angular4PaystackModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [Angular4PaystackComponent, Angular4PaystackDirective, Angular4PaystackEmbedComponent],
                    declarations: [Angular4PaystackComponent, Angular4PaystackDirective, Angular4PaystackEmbedComponent],
                    providers: [],
                }]
        }] });
